app/admin/resources/resource_factory.ts

```diff
@@ -0,0 +1,252 @@
+          type: "mixed",
+          isVisible: hideOnShow,
+          components: {
+            edit: "PhotoDropbox",
+          },
+        },
+        cover: {
+          isVisible: hideOnEdit,
+        },
+      };
+      newResource.options.actions = {
+        ...newResource.options.actions,
+        new: this.addNewUploadHook(newResource.options.actions?.new),
+        edit: this.addEditUploadHook(newResource.options.actions?.edit),
+        delete: this.addDeleteHook(newResource.options.actions?.delete),
+      };
+    }
+    return newResource;
+  }
+
+  private static addNewUploadHook(
+    additionalNewActions?: SubAction,
+  ): HookReturnValue {
+    return {
+      ...additionalNewActions,
+      before: async (request: ActionRequest): Promise<ActionRequest> => {
+        if (request.payload?.uploadPhoto !== undefined) {
+          request.payload = {
+            ...request.payload,
+            cover: await FilesService.uploadMultipartFile(
+              request.payload.uploadPhoto as MultipartFile,
+            ),
+          };
+          delete request.payload.uploadPhoto;
+        }
+        return request;
+      },
+    } as HookReturnValue;
+  }
+
+  private static addEditUploadHook(
+    additionalEditActions?: SubAction,
+  ): HookReturnValue {
+    return {
+      ...additionalEditActions,
+      before: async (request: ActionRequest): Promise<ActionRequest> => {
+        if (request.payload?.uploadPhoto !== undefined) {
+          if (typeof request.payload.cover === "string") {
+            request.payload = {
+              ...request.payload,
+              cover: await ResourceFactory.replaceExistingCover(
+                request.payload.cover,
+                request.payload.uploadPhoto as MultipartFile,
+              ),
+            };
+          } else {
+            request.payload = {
+              ...request.payload,
+              cover: await FilesService.uploadMultipartFile(
+                request.payload.uploadPhoto as MultipartFile,
+              ),
+            };
+          }
+          delete request.payload.uploadPhoto;
+        }
+        return request;
+      },
+    } as HookReturnValue;
+  }
+
+  private static addDeleteHook(
+    additionalDeleteActions?: SubAction,
+  ): HookReturnValue {
+    return {
+      ...additionalDeleteActions,
+      after: async (
+        entity: ImprovedRecordActionResponse,
+      ): Promise<RecordActionResponse> => {
+        const currentCover: ParamsTypeValue = entity.record.params.cover;
+        if (typeof currentCover === "string") {
+          await FilesService.deleteFileWithKey(currentCover);
+        }
+        return entity;
+      },
+    } as HookReturnValue;
+  }
+
+  private static replaceExistingCover(
+    currentCover: string,
+    newCover: MultipartFile,
+  ): Promise<string> {
+    return FilesService.deleteFileWithKey(currentCover).then(() => {
+      return FilesService.uploadMultipartFile(newCover);
+    });
+  }
```

Comment:
one of my linters (rightfully) suggests that this could be converted to an async function
also you probably should first upload then delete, so that you don't end up with no file if the upload fails or takes a long time

Reply:
actually thinking about this further: you probably should upload in the before hook and delete in the after hook